### PR TITLE
Maintain text color transforming from media to cover

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Related: 
- https://github.com/WordPress/gutenberg/pull/38458

## What?
Extends the Media & Text to Cover transform introduced in #38458 to apply any text color selection from the Media & Text block to the inner blocks moved to the new Cover block. 

## Why?
This aims to more faithfully keep user color selections when transforming from Media & Text blocks to Cover blocks.

## How?

- Applies any block support color selections from the Media & Text block to the new Cover's inner blocks
- Only applies the Media & Text's `textColor` or `style.color.text` value if the inner block doesn't already have its own color block support selections.

## Testing Instructions

1. Test transforming Media & Text block with named text and background color selections to a Cover block
    - The new Cover block's inner paragraph should have the named text color
    - Cover block should get the named background color
2. Transform Media & Text with gradient background and custom text color
    - New Cover block's inner paragraph should get the custom color
    - Cover block should have the gradient
3. Transform Media & Text with multiple different types of inner blocks. 
    - Ensure you have some inner blocks that do not opt into color block support. 
    - Make color selections on some inner blocks to ensure they are respected
    - After transforming, the inner blocks within the new Cover block should appear to have the same color selections as before
    
<details>
<summary>Some example block code pre transforms</summary>

```html
<!-- wp:media-text {"align":"","mediaId":246,"mediaLink":"http://localhost:4759/?attachment_id=246","mediaType":"image","backgroundColor":"primary","textColor":"background"} -->
<div class="wp-block-media-text is-stacked-on-mobile has-background-color has-primary-background-color has-text-color has-background"><figure class="wp-block-media-text__media"><img src="http://localhost:4759/wp-content/uploads/2022/04/Cano-Crystales-1080-1024x768.jpeg" alt="" class="wp-image-246 size-full"/></figure><div class="wp-block-media-text__content"><!-- wp:paragraph {"placeholder":"Content…"} -->
<p>This is the text part...</p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:media-text -->

<!-- wp:media-text {"mediaId":229,"mediaLink":"http://localhost:4759/?attachment_id=229","mediaType":"image","style":{"color":{"text":"#5b5252"}},"gradient":"electric-grass"} -->
<div class="wp-block-media-text alignwide is-stacked-on-mobile has-electric-grass-gradient-background has-text-color has-background" style="color:#5b5252"><figure class="wp-block-media-text__media"><img src="http://localhost:4759/wp-content/uploads/2022/04/robin-noguier-CZJp1S4bZos-unsplash-1024x683.jpg" alt="" class="wp-image-229 size-full"/></figure><div class="wp-block-media-text__content"><!-- wp:paragraph {"placeholder":"Content…"} -->
<p>More text. Parent has gradient background and custom text color.</p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:media-text -->

<!-- wp:media-text {"mediaId":249,"mediaLink":"http://localhost:4759/?attachment_id=249","mediaType":"image","style":{"color":{"text":"#de874a"}},"backgroundColor":"pale-cyan-blue"} -->
<div class="wp-block-media-text alignwide is-stacked-on-mobile has-pale-cyan-blue-background-color has-text-color has-background" style="color:#de874a"><figure class="wp-block-media-text__media"><img src="http://localhost:4759/wp-content/uploads/2022/04/Penol_1080-1024x768.jpeg" alt="" class="wp-image-249 size-full"/></figure><div class="wp-block-media-text__content"><!-- wp:paragraph {"placeholder":"Content…"} -->
<p>Lorem ipsum...</p>
<!-- /wp:paragraph -->

<!-- wp:buttons {"style":{"spacing":{"margin":{"top":"37px","bottom":"37px"}}}} -->
<div class="wp-block-buttons" style="margin-top:37px;margin-bottom:37px"><!-- wp:button {"backgroundColor":"primary","style":{"color":{"text":"#e98484"}}} -->
<div class="wp-block-button"><a class="wp-block-button__link has-primary-background-color has-text-color has-background" style="color:#e98484">Button</a></div>
<!-- /wp:button -->

<!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link">Button</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->

<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>Column 1 inherits its color.</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph {"textColor":"primary"} -->
<p class="has-primary-color has-text-color">Column 2 with its own color selection/</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->

<!-- wp:image {"id":249,"sizeSlug":"thumbnail","linkDestination":"none"} -->
<figure class="wp-block-image size-thumbnail"><img src="http://localhost:4759/wp-content/uploads/2022/04/Penol_1080-150x150.jpeg" alt="" class="wp-image-249"/><figcaption>Image in text portion of Media &amp; Text</figcaption></figure>
<!-- /wp:image --></div></div>
<!-- /wp:media-text -->
```

</details>

## Screenshots or screencast 

https://user-images.githubusercontent.com/60436221/165662522-c8a955b4-a90e-46e7-8fc2-0c4bec993da0.mp4

### CODE REVIEW

File: packages/block-library/src/media-text/transforms.js [hunk 1]
```diff
@@ -141,22 +142,74 @@ const transforms = {
...
+					const {
+						attributes: { style: innerStyle },
+					} = innerBlock;
+
+					// Only apply the media and text color if the inner block
+					// doesn't set its own color block support selection.
+					if (
+						innerBlock.attributes.textColor ||
+						innerStyle?.color?.text
+					) {
```
Comment (aaronrobertshaw) [inline packages/block-library/src/media-text/transforms.js:181]: We might also be able to check for color block support on the inner blocks.

**Comment by glendaviesnz** [comment]
Nice! ... a minor issue I can see is that a user can set the `Media & Text` text color with the parent blocks color supports, but then if they convert to Cover and back, then changes to the parent `Media & Text` block colors has no effect on the children as they now have their own color specified. With the original PR the color is lost, but the parent color selector still works as expected when you switch back.

It is a compromise either way. I don't have a strong opinion about which compromise will cause the least grief, but I suspect more people would be confused by the fact that the Media & Text text color selection no longer works for them than those that wondered why the text color vanished and didn't just apply it again via the parent - but purely a hunch on my part.

**Comment by aaronrobertshaw** [comment]
> a minor issue I can see is that a user can set the Media & Text text color with the parent blocks color supports, but then if they convert to Cover and back

Are users more likely to undo rather than re-transform back?

> It is a compromise either way.

I'd lean towards not losing user selections as the priority. 

The number transforming from Media and Text to Cover block would be higher than the number transforming and then transforming back (lessened still by those using undo). The number experiencing an issue with the color now being on the child block would be only the further subset that wish to tweak the color again after the fact.

My hunch is that the greater impact is the initial loss of color selection on the first transform.

**Comment by glendaviesnz** [comment]
> Are users more likely to undo rather than re-transform back?

🤷‍♂️  ... I am probably just a bit wary of doing anything that might add further complexity to the Cover block, but actually, this doesn't affect the Cover block in any way that would require future deprecations to unwind, so I am easy either way.

**Comment by apeatling** [comment]
> My hunch is that the greater impact is the initial loss of color selection on the first transform.

Agree.

**Comment by walbo** [comment]
Thanks for the additional PR. I do think it makes a better user experience to not loose the color. A bit confusing that it will change location if you transform back and forth, but I belive this is a compromise we will have to make as long as the cover block doesn't have text color on the block. The plus side is that we can change the behavor in the future without deprecations if the cover block adds it.

So a +1 from my side.

**Review by walbo — APPROVED** [review]

**Pull request merged**